pkg/resource/autonaming/pattern.go

```diff
@@ -0,0 +1,183 @@
+// Copyright 2016-2024, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package autonaming
+
+import (
+	"crypto"
+	"encoding/hex"
+	"fmt"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"github.com/gofrs/uuid"
+	"github.com/pulumi/pulumi/pkg/v3/backend"
+	"github.com/pulumi/pulumi/pkg/v3/backend/httpstate"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/resource/config"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/resource/urn"
+	"lukechampine.com/frand"
+)
+
+// stackPatternEval is a helper struct for resolving stack-level expressions in autonaming patterns.
+// It's used to resolve ${org}, ${project}, ${stack}, and ${config.key} expressions in patterns.
```

Comment:
I feel like this should be `${organization}` -- that'd match other things we have like `getStack`/`getOrganization`, and how we interpolate in YAML I believe?